### ISSUE TITLE
Use Modelica solver system

### ISSUE THREAD

**Issue by goatchurchprime** [NikraDAP/FreeCAD-NikraDAP#2]
I can't be sure, but it looks like you may be implementing your physics solver in Python, which is why the features are limited to planes and pendulums.

Have you looked at Modelica? This is an industrial grade open source physics solver that is used to simulate things as complicated as automobile drive trains. https://openmodelica.org/

The high level interface to Modelica is in the form of a networked block diagram GUI with a library of components, such these: https://doc.modelica.org/Modelica%204.0.0/Resources/helpDymola/Modelica_Mechanics_MultiBody_Examples.html#Modelica.Mechanics.MultiBody.Examples

This graphical block network interface outputs a Modelica language file, which is a simple object oriented language described here: https://en.wikipedia.org/wiki/Modelica#Examples

The Modelican language file is then compiled and numerically integrated in time to produce the simulation.  

If your FreeCAD interface was able to generate Modelica files directly and read the simulated results, then the physics solver problem would be done.

**Comment by NikraDAP** [comment]
I appreciate your comment.
The functionality of NikraDAP is by no means limited to pendulums.  It is just the simplest option to debug, and the simplest to fit the description into 20 mins.  There are quite a comprehensive list of force and joint options under the NikraDAP hood, which will magically appear once they are fully debugged in the near future.  (You can remove the commented out lines at the beginning of DapTools, if you wish to play with them)
As far as python is concerned, the heavy lifting is done by NumPy, which in turn is just a wrapper for real number crunchers like BLAS.  And I believe that is where most number crunching ends up anyway in all the fancy packages out there.
I believe (at least from my perspective) the basic philosophy of NikraDAP is less to be this super-powerful tool which will solve all the bells and whistles, and more to be a tool which is fully functional in its realm, but is nevertheless approachable to the average student - even those who wish to modify the code to add their own specific functionality.  (For this reason, I would guess that on average, every 2 lines of code, have their own more verbose explanation, specifically for the adventurous student).  Even though I have spent much of my career in the C-world, I agreed with others in our group, (and with the FreeCAD developers), and switched to python, as being more approachable to a wider user base.  And the same philosophy would suggest that, at least at this stage, simple is elegant in NikraDAP.
Of course, that doesn't preclude anyone else from developing their own Modelica interface for FreeCAD.  What beauty resides in the open source concept!

**Comment by cecilchurms** [comment]
P.S.  Permission was explicitly granted by P.E. Nikravesh to use his algorithm, which is also the text-book for the applicable course at Univ. Pretoria.  It was also natural to go his route - at least for the first while, until it might happen that Modelica turns out to be very easily integrated.   Maybe you will see "ModeliNikraDAP" in the future.

**Comment by ljplessis** [comment]
@goatchurchprime - thank you for the reference to Modelica - I will definately check it out.  @cecilchurms  that you also for the explanation on why we went the planar route and for reassuring @goatchurchprime that all planar mechanisms are not pendulums.  In fact any 3D / spacial multibody mechanism can be represented by Prov Nikravesh's algorithm (that is now implemented in NikraDAP) as long as the individual bodies (of the 3D / spacial mechanism) all move in a plane OR parallel planes.
